Add ASP.NET support

halter73 opened the pull request:
There's a lot left to do before undrafting, but I said I'd open a draft PR today!

I want to have `#!aspnet` enable the ASP.NET environment I set up in UseAspNet(). I also want to move this out of the C# Kernel so it can be used with other languages.

I'm still experimenting heavily with what the inner loop should look like, but you can get a feel for the current experience by looking at https://github.com/halter73/AspNetNotebooks/blob/master/AspNetMin.ipynb.

![image](https://user-images.githubusercontent.com/54385/99867641-f82d1380-2b6f-11eb-8692-98481d01efd7.png)


In my opinion, the really cool thing about this is being able to modify an ASP.NET web app, make a request, and see the response along with any diagnostic output in a fraction of a second without leaving the notebook. This is possible because we don't have to restart the process, host or server to deploy changes. I might do a quick video demo once this is a little more polished.

jonsequitur commented on src/Microsoft.DotNet.Interactive.CSharp/AspNetKernelExtensions.cs:
```diff
@@ -0,0 +1,169 @@
+
+using Microsoft.DotNet.Interactive.Commands;
+
+namespace Microsoft.DotNet.Interactive.CSharp
```
I suspect that this code will be more at home in the `Microsoft.DotNet.Interactive.Http` project.

jonsequitur commented on src/Microsoft.DotNet.Interactive.CSharp/AspNetKernelExtensions.cs:
```diff
@@ -0,0 +1,169 @@
+            var command = new SubmitCode(@"
+Environment.SetEnvironmentVariable($""ASPNETCORE_{WebHostDefaults.PreventHostingStartupKey}"", ""true"");
+
+class LogLevelMonitor : IOptionsMonitor<LoggerFilterOptions>
```
We can likely implement these things in library code, possibly as an `IKernelExtension`, rather than using a code submission.

colombod commented on src/Microsoft.DotNet.Interactive.CSharp/AspNetKernelExtensions.cs:
```diff
@@ -0,0 +1,169 @@
+        loggingBuilder.AddSimpleConsole(options => options.ColorBehavior = LoggerColorBehavior.Disabled);
+    });
+
+var __AspNet_HostRunAsyncTask = __AspNet_HostBuilder.Build().RunAsync();
```
Is there a way to modify a host that is already running?

halter73 replied:
Not yet, but I would love to add support for that.

halter73 commented on src/Microsoft.DotNet.Interactive.Http/AspNetKernelExtensions.cs:
```diff
@@ -0,0 +1,323 @@
+                    // We would do an "is" check on the kernel type, but we don't have a reference to the containing project
+                    if (kernel.Name == "csharp")
+                    {
+                        await kernel.SendAsync(new SubmitCode($"using {typeof(EndpointRouteBuilderExtensions).Namespace};"), CancellationToken.None);
```
Add all usings here instead of to the CSharpKernel by default.

halter73 commented on src/dotnet-interactive/CommandLine/CommandLineParser.cs:
```diff
@@ -479,15 +480,17 @@ static HttpPortRange ParsePortRangeOption(ArgumentResult result)
                     .UseMathAndLaTeX()
-                    .UseDotNetVariableSharing(),
+                    .UseDotNetVariableSharing()
+                    .UseAspNet(),
```
Only use on C# kernel for now.

halter73 commented on src/Microsoft.DotNet.Interactive.Http/AspNetKernelExtensions.cs:
```diff
@@ -0,0 +1,323 @@
+        private const string _aspnetLogsKey = "aspnet-logs";
+
+        public static T UseAspNet<T>(this T kernel)
+            where T : DotNetKernel
```
For first round, limit to CSharpKernel.

halter73 commented on src/Microsoft.DotNet.Interactive.Http/AspNetKernelExtensions.cs:
```diff
@@ -0,0 +1,323 @@
+            }, HtmlFormatter.MimeType);
+
+            //KernelInvocationContext.Current.Publish(...)
+            //kernel.AddMiddleware()
```
Remove dead code.

halter73 commented on src/Microsoft.DotNet.Interactive.Http/AspNetKernelExtensions.cs:
```diff
@@ -0,0 +1,323 @@
+            {
+                Handler = CommandHandler.Create<string, string, KernelInvocationContext>(async (from, name, context) =>
+                {
+                    // REVIEW: Commands cannot run concurrently, right?
```
right

halter73 commented on src/Microsoft.DotNet.Interactive.FSharp/Microsoft.DotNet.Interactive.FSharp.fsproj:
```diff
@@ -50,4 +50,8 @@
 
+  <ItemGroup>
+    <FrameworkReference Include="Microsoft.AspNetCore.App" />
+  </ItemGroup>
```
Revert all kernel csproj changes

jonsequitur commented on src/Microsoft.DotNet.Interactive.FSharp/FSharpKernel.fs:
```diff
@@ -373,7 +373,7 @@ type FSharpKernel () as this =
             false
 
-    override _.SetVariableAsync(name: string, value: Object) : Task = 
+    override _.SetVariableAsync(name: string, value: Object, declaredType: Type) : Task = 
```
@TIHan This is an interesting and useful change to the `SetVariableAsync` that might should work its way down to FSI. Thoughts?

halter73 commented:
![image](https://user-images.githubusercontent.com/54385/102668032-c9d72f80-413f-11eb-952e-8cc8a95e2f8e.png)

I think this is finally ready to merge! I've updated the PR to minimize the impact on the kernels themselves. This does add an optional declared Type parameter to DotNetKernel.SetVariableAsync(), but it shouldn't cause any problems for people taking dependencies on the Kernels like before when I changed the C# kernel's TFM from netstandard2.1 to net5.0.

halter73 commented on src/Microsoft.DotNet.Interactive.AspNetCore/AspNetCoreKernelExtensions.cs:
```diff
@@ -0,0 +1,104 @@
+        }
+
+        // REVIEW: Should we make this a public CSharp/DotNetKernel extension method?
+        private static Task SetVariableAsync<T>(this CSharpKernel kernel, string name, T value)
```
Should we make this public?

jonsequitur replied:
I think a public version of this is useful but should be on `DotNetKernel` (or ` T where T : DotNetKernel`) rather than on `CSharpKernel`.

jonsequitur commented on src/Microsoft.DotNet.Interactive.AspNetCore/HttpClientFormatter.cs:
```diff
@@ -0,0 +1,129 @@
+                div(h2("Request"), hr(), requestLine, requestHeaders, requestBody),
+                div(h2("Response"), hr(), responseLine, responseHeaders, responseBody));
+
+            //Pocket.LoggerExtensions.Info(Pocket.Logger.Log, output.ToString());
```
Want to remove this comment?

halter73 commented:
⬆️📅 I addressed the PR feedback and added more tests.

halter73 commented:
@jonsequitur The PR builds are consistently hanging after printing `Running tests: D:\a\1\s\artifacts\bin\Microsoft.DotNet.Interactive.Tests\Release\net5.0\Microsoft.DotNet.Interactive.Tests.dll [net5.0|x64]` This is the same on Windows and Linux. Locally, when I run `dotnet test -c Release` in `src\dotnet-interactive.Tests` which appears to be the test project behind Microsoft.DotNet.Interactive.Tests.dll, all 120 tests pass in 24 seconds. Any idea what's going on?

jonsequitur commented:
We've been seeing this issue in our CI builds for a few weeks. It's not related to your PR.

Have you rebased recently?

halter73 commented:
@jonsequitur In my last commit, I added `#!aspnet-stop` so I could call it from AspNetCoreTests.DisposeAsync() and that looks to have fixed the issues with the test project hanging on AzDO. Should I merge this now?
